torch/csrc/jit/script/final_returns.cpp

```diff
@@ -1,106 +1,288 @@
+    for (auto it = nodes.begin(); it != nodes.end() && *it != return_node;) {
+      auto node = it;
+      it++;
+      if (*node != block->return_node()) {
+        node->destroy();
+      }
+    }
+    if (return_node->kind() == prim::ReturnStmt) {
+      return_node->destroy();
+    }
+  }
+
+  void checkNoLoopReturn(Block* block) {
+    for (Node* n : block->nodes()) {
+      if (n->kind() == prim::ReturnStmt) {
+        throw ErrorReport(n->getSourceLocation())
+            << "Return statements within loops are not yet supported\n";
+      }
+      for (Block* b : n->blocks()) {
+        checkNoLoopReturn(b);
+      }
+    }
+  }
+
+  void makeReturnsFinal(Block* block) {
+    auto ret_status = WONT_RETURN;
+    for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+      Node* node = *it;
+      it++;
+      switch (node->kind()) {
+        case prim::ReturnStmt: {
+          ret_status = WILL_RETURN;
+          block_return_vals[block] = node->input();
+        } break;
+        case prim::If: {
+          ret_status = handleIf(node);
+        } break;
+        case prim::Loop:
+          checkNoLoopReturn(node->blocks().at(0)); // not yet supported
+        default:
           break;
-        }
-        // (2) all branches return
-        if (true_final.returns_ && false_final.returns_) {
-          changed.emplace_back(
-              if_stmt.withNewBranches(true_final.stmts_, false_final.stmts_));
-          return {true, List<Stmt>::unsafeCreate(range, std::move(changed))};
-        }
-        throw ErrorReport(if_stmt)
-            << "This if statement contains some paths that return and some paths that do not. "
-            << "If statements must either entirely return or never return.";
-      } break;
-      case TK_WHILE:
-      case TK_FOR:
-        changed.emplace_back(stmt);
-        checkNoReturn(stmt);
+      }
+      if (ret_status == WILL_RETURN) {
+        deleteAfterReturnNodes(block, node);
         break;
-      case TK_RETURN:
-        changed.emplace_back(stmt);
-        // ignore the rest the the block, it is dead.
-        return {true, List<Stmt>::unsafeCreate(range, std::move(changed))};
-      default:
-        changed.emplace_back(stmt);
+      } else if (ret_status == MIGHT_RETURN) {
+        if (it != block->nodes().end()) {
+          ret_status = guardBlockNodes(
+              block, getSentinelVal(node), getReturnVal(node), it);
+        }
         break;
+      }
+    }
+    if (block_status.count(block) == 0) {
+      block_status[block] = ret_status;
+    } else {
+      // Guarded return blocks have their status set prior
+      AT_ASSERT(
+          block_status[block] == WILL_RETURN &&
+          block->nodes().begin() == block->nodes().end());
     }
   }
-  if (return_none) {
-    // add an implicit return none node
-    changed.emplace_back(
-        Return::create(range, Expr(Compound::create(TK_NONE, range, {}))));
+
+  Value* getBottomVal() {
```

Comment:
If the compiler universinoally created and passed this in, you could avoid having to repeat this whole thing here.